url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/builders/CronTriggerBuilder.java

```diff
@@ -0,0 +1,176 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.quartz.builders;
+
+import java.text.ParseException;
+import java.util.Date;
+import java.util.TimeZone;
+
+import org.quartz.jobs.JobDataMap;
+import org.quartz.triggers.CronExpression;
+import org.quartz.triggers.CronTrigger;
+import org.quartz.triggers.CronTriggerImpl;
+import org.quartz.triggers.OperableTrigger;
+
+/**
+ * <code>CronScheduleBuilder</code> is a {@link ScheduleBuilder} that defines {@link CronExpression}-based schedules for <code>Trigger</code>s.
+ * <p>
+ * Quartz provides a builder-style API for constructing scheduling-related entities via a Domain-Specific Language (DSL). The DSL can best be utilized
+ * through the usage of static imports of the methods on the classes <code>TriggerBuilder</code>, <code>JobBuilder</code>, <code>DateBuilder</code>,
+ * <code>JobKey</code>, <code>TriggerKey</code> and the various <code>ScheduleBuilder</code> implementations.
+ * </p>
+ * <p>
+ * Client code can then use the DSL to write code such as this:
+ * </p>
+ *
+ * <pre>
+ *
+ * JobDetail job = newJob(MyJob.class).withIdentity(&quot;myJob&quot;).build();
+ * 
+ * Trigger trigger = newTrigger().withIdentity(triggerKey(&quot;myTrigger&quot;, &quot;myTriggerGroup&quot;))
+ *     .withSchedule(simpleSchedule().withIntervalInHours(1).repeatForever()).startAt(futureDate(10, MINUTES)).build();
+ * 
+ * scheduler.scheduleJob(job, trigger);
+ *
+ * </pre>
+ */
+public class CronTriggerBuilder extends TriggerBuilder {
+
+  private String cronExpression;
+  private TimeZone tz = null;
+  private int misfireInstruction = CronTrigger.MISFIRE_INSTRUCTION_SMART_POLICY;
+
+  /**
+   * Constructor
+   *
+   * @param cronExpression
+   */
+  private CronTriggerBuilder(String cronExpression) {
+
+    this.cronExpression = cronExpression;
+  }
+
+  /**
+   * Create a CronScheduleBuilder with the given cron-expression.
+   *
+   * @param cronExpression the cron expression to base the schedule on.
+   * @return the new CronScheduleBuilder
+   * @throws ParseException
+   * @see CronExpression
+   */
+  public static CronTriggerBuilder cronTriggerBuilder(String cronExpression) throws ParseException {
+
+    CronExpression.validateExpression(cronExpression);
+    return new CronTriggerBuilder(cronExpression);
+  }
+
+  /**
+   * Build the actual Trigger -- NOT intended to be invoked by end users, but will rather be invoked by a TriggerBuilder which this ScheduleBuilder is
+   * given to.
+   */
+  @Override
+  public OperableTrigger instantiate() {
+
+    CronTriggerImpl ct = new CronTriggerImpl();
+
+    try {
+      ct.setCronExpression(cronExpression);
+    } catch (ParseException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)